# Lab 07 - Bài tập

In [ ]:
import numpy as np
import scipy.stats as st
import math

## Luyện tập

### Bài 1

Trong thập niên 80, trọng lượng trung bình của thanh niên là 48 kg. Nay để xác định lại trọng lượng ấy, người ta chọn ngẫu nhiên 100 thanh niên đo trọng lượng trung bình là 50 kg và phương sai mẫu $s^2 = (10 kg)^2$. Thử xem trọng lượng thanh niên hiện nay phải chăng có thay đổi, với mức ý nghĩa là 1%?

   _Đáp án. Trọng lượng thanh niên hiện nay không thay đổi so với trước kia._

In [ ]:
n = 100
s2 = 10 ** 2

x_bar = 50
alpha = 0.01

**B1**: Xây dựng giả thuyết kiểm định

$$H_o: \mu = 48$$

$$H_a: \mu \neq 48$$

**B2** Mức ý nghĩa của kiểm định $\alpha = 0.01$

**B3** Tính giá trị kiểm định từ mẫu
$$t_0 = \frac{\bar{x} - \mu_0}{\frac{s}{\sqrt{n}}} \approx 50$$

In [ ]:
mu = 48
z = (x_bar - mu) / (math.sqrt(s2) / math.sqrt(n))
print('- Trị kiểm định: ', z)

- Trị kiểm định:  2.0


**B4** Xác định `p-value` hoặc `miền bác  bỏ` (critical region) của kiểm định

- Trường hợp xác định miền bác bỏ

$$\{ t: |t| > t^{n-1}_{\alpha/2} \}$$

$$\{ t: |t| > 2 \}$$

In [ ]:
p_value = 2 * (1 - st.norm.cdf(abs(z)))
print('- p-value: ', p_value)

- p-value:  0.04550026389635842


**B5** Đưa ra quyết định về giả thuyết vô hiệu. Rút ra kết luận chung Kết luận

- Không bác bỏ giả thuyết $H_0$ (vì trị kiểm định không thuộc miền bác bỏ HOẶC p-value < $\alpha$)
- Vậy cân nặng của thanh niên không có sự thay đổi.

### Bài 2

Một cửa hàng thực phẩm nhận thấy thời gian vừa qua trung bình một khách hàng mua 25 ngàn đồng thực phẩm trong ngày. Nay cửa hàng chọn ngẫu nhiên 15 khách hàng thấy trung bình một khách hàng mua 24 ngàn đồng trong ngày và phương sai mẫu là $s^2$ =  (2 ngàn đồng $)^2$

   Với mức ý nghĩa là 5%, kiểm định xem có phải sức mua của khách hàng hiện nay thực sự giảm sút hay không. Biết rằng sức mua của khách hàng có phân phối chuẩn.

   _Đáp án. Sức mua của khách hàng hiện nay thực sự giảm sút_

In [ ]:
mu = 25
n = 15
s = 2
x_bar = 24
alpha = 0.05

**B1**: Xây dựng giả thuyết kiểm định

$$H_o: \mu \le 25$$

$$H_a: \mu \gt 25$$

**B2** Mức ý nghĩa của kiểm định $\alpha = 0.05$

**B3** Tính giá trị kiểm định từ mẫu
$$t_0 = \frac{\bar{x} - \mu_0}{\frac{s}{\sqrt{n}}} \approx 24$$

In [ ]:
t = (x_bar - mu) / (s / math.sqrt(n))
print('- Trị kiểm định: ', t)

- Trị kiểm định:  -1.9364916731037087


**B4** Xác định `p-value` hoặc `miền bác  bỏ` (critical region) của kiểm định

- Trường hợp xác định miền bác bỏ

$$\{ t: |t| > t^{n-1}_{\alpha/2} \}$$

$$\{ t: |t| > -1.9364916731037087 \}$$

In [ ]:
p_value = 1 - st.t.cdf(t, n - 1)
print('- p-value: ', p_value)

- p-value:  0.04550026389635842


**B5** Đưa ra quyết định về giả thuyết vô hiệu. Rút ra kết luận chung Kết luận

- Không bác bỏ giả thuyết $H_0$ (vì trị kiểm định không thuộc miền bác bỏ HOẶC p-value < $\alpha$)
- Vậy sức mua của khách hàng hiện nay thực sự giảm sút.

### Bài 3

Một máy sản suất tự động, lúc đầu tỷ lệ sản phẩm loại A là 20%. Sau khi áp dụng một phương pháp cải tiến sản xuất mới, người ta lấy 40 mẫu, mỗi mẫu gồm 10 sản phẩm đề kiểm tra. Kết quả kiểm tra cho ở bảng sau:

| **Số sản phẩm loại A** |  1   |  2   |  3   |  4   |  5   |  6   |  7   |  8   |  9   |  10  |
| ---------------------- | :--: | :--: | :--: | :--: | :--: | :--: | :--: | :--: | :--: | :--: |
| **Số mẫu**             |  2   |  0   |  4   |  6   |  8   |  10  |  4   |  5   |  1   |  0   |

 Với mức ý nghĩa 5%. Hãy cho kết luận về phương pháp sản suất này.

 _Đáp án.Phương pháp cải tiến sản xuất mới thay đổi tỉ lệ sản phẩm loại A_

In [ ]:

# Dữ liệu
p_0 = 0.2  # Tỷ lệ giả thuyết
n = 400    # Tổng số sản phẩm (40 mẫu x 10 sản phẩm)
x = 155    # Tổng số sản phẩm loại A trong 40 mẫu

# Tỷ lệ mẫu
p_hat = x / n

# Tính Z
z = (p_hat - p_0) / math.sqrt(p_0 * (1 - p_0) / n)

# Tính P-value từ CDF
p_value = 2 * (1 - st.norm.cdf(abs(z)))  # Nhân đôi vì kiểm định hai đuôi

# In kết quả
print(f"Giá trị Z tính được: {z:.4f}")
print(f"P-value: {p_value:.4f}")

# Kiểm định
alpha = 0.05
if p_value < alpha:
    print("Bác bỏ giả thuyết H₀: Phương pháp cải tiến đã thay đổi tỷ lệ sản phẩm loại A.")
else:
    print("Không bác bỏ giả thuyết H₀: Tỷ lệ sản phẩm loại A không thay đổi.")


Giá trị Z tính được: 9.3750
P-value: 0.0000
Bác bỏ giả thuyết H₀: Phương pháp cải tiến đã thay đổi tỷ lệ sản phẩm loại A.


### Bài 4

Tỷ lệ phế phẩm của một nhà máy trước đây là 5%. Năm nay nhà máy áp dụng một biện pháp kỹ thuật mới. Để nghiên cứu tác dụng của biện pháp kỹ thuật mới, người ta lấy một mẫu gồm 800 sản phẩm để kiểm tra và thấy có 24 phế phẩm.

a. Với α = 0.01. Hãy cho kết luận về biện pháp kỹ thuật mới này?

b. Nếu nhà máy báo cáo tỷ lệ phế phẩm sau khi áp dụng biện pháp kỹ thuật mới là
2% thì có chấp nhận được không? (α = 0.01).

_Đáp án. (a) Biện pháp kĩ thuật mới làm thay đổi tỉ lệ phế phẩm (b) Nhà máy báo cáo tỷ lệ phế phẩm là chấp nhận được._

In [ ]:
import math
import scipy.stats as stats

# Dữ liệu
n = 800    # Tổng số sản phẩm
x = 24     # Số phế phẩm
alpha = 0.01  # Mức ý nghĩa

# Tỷ lệ phế phẩm trong mẫu
p_hat = x / n

# a) Kiểm định giả thuyết với p0 = 0.05
p_0 = 0.05  # Tỷ lệ phế phẩm giả thuyết

# Tính Z
z = (p_hat - p_0) / math.sqrt(p_0 * (1 - p_0) / n)

# Tính P-value từ CDF
p_value_a = 2 * (1 - stats.norm.cdf(abs(z)))  # Nhân đôi vì kiểm định hai đuôi

# In kết quả bài (a)
print(f"Giá trị Z bài a: {z:.4f}")
print(f"P-value bài a: {p_value_a:.4f}")

# Kết luận bài (a)
if p_value_a < alpha:
    print("Bài a: Bác bỏ giả thuyết H₀: Biện pháp kỹ thuật mới làm thay đổi tỷ lệ phế phẩm.")
else:
    print("Bài a: Không bác bỏ giả thuyết H₀: Biện pháp kỹ thuật mới không làm thay đổi tỷ lệ phế phẩm.")

# b) Kiểm định giả thuyết với p0 = 0.02
p_0_b = 0.02  # Tỷ lệ phế phẩm giả thuyết mới

# Tính Z cho bài (b)
z_b = (p_hat - p_0_b) / math.sqrt(p_0_b * (1 - p_0_b) / n)

# Tính P-value từ CDF cho bài (b)
p_value_b = 2 * (1 - stats.norm.cdf(abs(z_b)))  # Nhân đôi vì kiểm định hai đuôi

# In kết quả bài (b)
print(f"Giá trị Z bài b: {z_b:.4f}")
print(f"P-value bài b: {p_value_b:.4f}")

# Kết luận bài (b)
if p_value_b < alpha:
    print("Bài b: Bác bỏ giả thuyết H₀: Tỷ lệ phế phẩm sau biện pháp kỹ thuật không phải 2%.")
else:
    print("Bài b: Không bác bỏ giả thuyết H₀: Tỷ lệ phế phẩm sau biện pháp kỹ thuật là chấp nhận được.")


Giá trị Z bài a: -2.5955
P-value bài a: 0.0094
Bài a: Bác bỏ giả thuyết H₀: Biện pháp kỹ thuật mới làm thay đổi tỷ lệ phế phẩm.
Giá trị Z bài b: 2.0203
P-value bài b: 0.0434
Bài b: Không bác bỏ giả thuyết H₀: Tỷ lệ phế phẩm sau biện pháp kỹ thuật là chấp nhận được.


## Kiểm định đặc trưng tỉ lệ (p)

### Bài 5

Dataset 18 (__M&M__) chứa thông tin trọng lượng và màu sắc của mẫu các viên kẹo của hãng M&M.

Sử dụng mẫu trên bạn hãy kiểm tra phát biểu sau: __"20% kẹo M&M có màu đỏ"__ với mức ý nghĩa là 5%

*Hướng dẫn: tính tỷ lệ các viên kẹo có màu đỏ từ dữ liệu, sử dụng kết quả để kiểm định phát biểu trên*

In [ ]:
import pandas as pd

# Đọc dữ liệu từ file CSV (thay đổi đường dẫn tới file của bạn)
data = pd.read_excel("./18_M_M.xls")

data_red = data['Red']

# Tổng số mẫu
n = len(data_red)

# Tỷ lệ viên kẹo màu đỏ trong mẫu
p_hat = np.mean(data_red)

# Giả thuyết tỷ lệ viên kẹo màu đỏ là 20% (p0 = 0.2)
p_0 = 0.2

# Tính Z-score
z = (p_hat - p_0) / np.sqrt((p_0 * (1 - p_0)) / n)

# Tính P-value từ CDF (kiểm định hai đuôi)
p_value = 2 * (1 - stats.norm.cdf(abs(z)))

# In kết quả
print(f"Tỷ lệ viên kẹo màu đỏ trong mẫu: {p_hat:.4f}")
print(f"Giá trị Z: {z:.4f}")
print(f"P-value: {p_value:.4f}")

# Mức ý nghĩa
alpha = 0.05

# So sánh P-value với mức ý nghĩa
if p_value < alpha:
    print("Bác bỏ giả thuyết H₀: Tỷ lệ viên kẹo màu đỏ không phải là 20%.")
else:
    print("Không bác bỏ giả thuyết H₀: Tỷ lệ viên kẹo màu đỏ là 20%.")

Tỷ lệ viên kẹo màu đỏ trong mẫu: 0.8635
Giá trị Z: 8.6196
P-value: 0.0000
Bác bỏ giả thuyết H₀: Tỷ lệ viên kẹo màu đỏ không phải là 20%.


### Bài 6

Dataset 3 (__FRESH15__) chứa thông tin khảo sát những thay đổi về trọng lượng (weight - WT) và chỉ số khối của cơ thể (BMI) của sinh viên năm nhất tại một trường Đại Học. Thời điểm khảo sát là bắt đầu học kỳ (tháng 9 - SEP) và kết thúc học kỳ (tháng 4 - APR), các sinh viên được chọn ngẫu nhiên trong trường.

Sử dụng mẫu trên, hãy kiểm tra phát biểu:__"50% số sinh viên trong trường là nam"__ với mức ý nghĩa 50%

Các thuộc tính trong bộ dữ liệu:
- SEX: Giới tính
- WTSEP: trọng lượng đo vào tháng 9 (mới nhập học)
- WTAPR: trọng lượng đo vào tháng 4 (kết thúc năm học)
- BMISP: BMI tháng 9
- BMIAP: BMI tháng 4

In [ ]:
import pandas as pd
import scipy.stats as stats

# Đọc dữ liệu từ file Excel
df = pd.read_excel('./03_FRESH15.xls')

# Kiểm tra các giá trị có trong cột 'SEX' để đảm bảo có các giá trị 'M' và 'F'
print(df['SEX'].value_counts())

# Tính tỷ lệ sinh viên nam trong mẫu
num_male = (df['SEX'] == 'M').sum()
n = len(df)  # Tổng số sinh viên trong mẫu
p_hat = num_male / n  # Tỷ lệ sinh viên nam trong mẫu

# Giả thuyết tỷ lệ sinh viên nam là 50% (p0 = 0.5)
p_0 = 0.5

# Tính Z-score
z = (p_hat - p_0) / ( (p_0 * (1 - p_0)) / n ) ** 0.5

# Tính P-value từ CDF (kiểm định hai đuôi)
p_value = 2 * (1 - stats.norm.cdf(abs(z)))

# In kết quả
print(f"Tỷ lệ sinh viên nam trong mẫu: {p_hat:.4f}")
print(f"Giá trị Z: {z:.4f}")
print(f"P-value: {p_value:.4f}")

# Mức ý nghĩa
alpha = 0.05

# So sánh P-value với mức ý nghĩa
if p_value < alpha:
    print("Bác bỏ giả thuyết H₀: Tỷ lệ sinh viên nam không phải là 50%.")
else:
    print("Không bác bỏ giả thuyết H₀: Tỷ lệ sinh viên nam là 50%.")


SEX
F    35
M    32
Name: count, dtype: int64
Tỷ lệ sinh viên nam trong mẫu: 0.4776
Giá trị Z: -0.3665
P-value: 0.7140
Không bác bỏ giả thuyết H₀: Tỷ lệ sinh viên nam là 50%.


## Kiểm định đặc trưng kỳ vọng ($\mu$) biết ($\sigma$)

### Bài 7

Dataset 13 __(VOLTAGE)__ liệt kê các mức điện áp được đo tại các hộ gia đình bởi các máy đo khác nhau: tại nhà (HOME), tại máy phát độc lập (Generator), tại trạm phát (UPS).

Công ty cung cấp điện cho vùng này cam kết rằng: __"Mục tiêu cung cấp điện cho mỗi hộ gia đình là 120V"__ . Biết rằng độ lệch chuẩn của tất cả các mức điện áp đo được là 0.24V.

Với mức ý nghĩa 1%. Bạn hãy kiểm tra xem công ty có làm đúng cam kết hay không?

In [ ]:
import pandas as pd
import scipy.stats as stats

# Đọc dữ liệu từ file Excel
df = pd.read_excel('./13_VOLTAGE.xls')

# Tính giá trị trung bình của cột 'Home' (mức điện áp tại các hộ gia đình)
mean_home = df['Home'].mean()

# Thông số từ bài toán
mu = 120  # Mức cam kết của công ty
sigma = 0.24  # Độ lệch chuẩn
n = len(df)  # Số ngày đo (số mẫu)

# Tính Z-score
z = (mean_home - mu) / (sigma / (n ** 0.5))

# Tính P-value từ Z-score (kiểm định hai đuôi)
p_value = 2 * (1 - stats.norm.cdf(abs(z)))

# In kết quả
print(f"Trung bình mức điện áp tại các hộ gia đình: {mean_home:.4f}")
print(f"Giá trị Z: {z:.4f}")
print(f"P-value: {p_value:.4f}")

# Mức ý nghĩa
alpha = 0.01

# So sánh P-value với mức ý nghĩa
if p_value < alpha:
    print("Bác bỏ giả thuyết H₀: Mức điện áp không phải là 120V.")
else:
    print("Không bác bỏ giả thuyết H₀: Mức điện áp là 120V.")


Trung bình mức điện áp tại các hộ gia đình: 123.6625
Giá trị Z: 96.5153
P-value: 0.0000
Bác bỏ giả thuyết H₀: Mức điện áp không phải là 120V.


## Kiểm định đặc trưng kỳ vọng ($\mu$) không biết ($\sigma$)

### Bài 8

Dataset 19(__SCREWS__) chứa thông tin khảo sát chiều dài một số đinh ốc được cung cấp bởi công ty Crown Bolt (đơn vị inches)

Trên bao bì của các đinh ốc do công ty sản xuất có ghi chiều dài của các đinh ốc là 3/4 inches. Sử dụng mẫu trên bạn hãy cho biết chiều dài của đinh ốc lấy từ mẫu thu nhận được từ dữ liệu trên có nhất quán với chiều dài ghi trên bao bì hay không? Kiểm định với mức ý nghĩa 5%

In [ ]:
import pandas as pd
import scipy.stats as stats

# Đọc dữ liệu từ file Excel
df = pd.read_excel('./19_SCREWS.xls', header=None)

# Tính giá trị trung bình và độ lệch chuẩn của mẫu
mean_screws = df[0].mean()
std_screws = df[0].std()
n = len(df)  # Số lượng quan sát (mẫu)

# Mức chiều dài theo bao bì
mu = 0.75

# Tính giá trị t
t = (mean_screws - mu) / (std_screws / (n ** 0.5))

# Tính P-value từ giá trị t (kiểm định hai đuôi)
p_value = 2 * (1 - stats.t.cdf(abs(t), df=n-1))

# In kết quả
print(f"Trung bình chiều dài đinh ốc trong mẫu: {mean_screws:.4f}")
print(f"Độ lệch chuẩn của mẫu: {std_screws:.4f}")
print(f"Giá trị t: {t:.4f}")
print(f"P-value: {p_value:.4f}")

# Mức ý nghĩa
alpha = 0.05

# So sánh P-value với mức ý nghĩa
if p_value < alpha:
    print("Bác bỏ giả thuyết H₀: Chiều dài đinh ốc không phải là 0.75 inches.")
else:
    print("Không bác bỏ giả thuyết H₀: Chiều dài đinh ốc là 0.75 inches.")


Trung bình chiều dài đinh ốc trong mẫu: 0.7468
Độ lệch chuẩn của mẫu: 0.0123
Giá trị t: -1.8249
P-value: 0.0741
Không bác bỏ giả thuyết H₀: Chiều dài đinh ốc là 0.75 inches.


### Bài 9

Dataset 2 __(BODYTEMP)__ chứa thông tin khảo sát nhiệt độ cơ thể từ một nhóm người vào cùng một thời điểm lúc 12h trưa.

Có một phát biểu nói rằng, nhiệt độ cơ thể người đo lúc 12h trưa trong quần thể này là 98,6°

Dựa vào mẫu trên, bạn hãy cho biết phát biểu trên đúng hay sai ở mức ý nghĩa là 5%.

In [ ]:
import pandas as pd
import scipy.stats as stats

# Đọc dữ liệu từ file Excel
df = pd.read_excel('./02_BODYTEMP.xls', header=None)

# Tính giá trị trung bình và độ lệch chuẩn của mẫu
mean_temp = df[0].mean()
std_temp = df[0].std()
n = len(df)  # Số lượng quan sát (mẫu)

# Nhiệt độ theo phát biểu
mu = 98.6

# Tính giá trị t
t_stat = (mean_temp - mu) / (std_temp / (n ** 0.5))

# Tính P-value từ giá trị t (kiểm định hai đuôi)
p_value = 2 * (1 - stats.t.cdf(abs(t_stat), df=n-1))

# In kết quả
print(f"Trung bình nhiệt độ cơ thể trong mẫu: {mean_temp:.4f}")
print(f"Độ lệch chuẩn của mẫu: {std_temp:.4f}")
print(f"Giá trị t: {t_stat:.4f}")
print(f"P-value: {p_value:.4f}")

# Mức ý nghĩa
alpha = 0.05

# So sánh P-value với mức ý nghĩa
if p_value < alpha:
    print("Bác bỏ giả thuyết H₀: Nhiệt độ cơ thể không phải là 98.6°.")
else:
    print("Không bác bỏ giả thuyết H₀: Nhiệt độ cơ thể là 98.6°.")


Trung bình nhiệt độ cơ thể trong mẫu: 98.2000
Độ lệch chuẩn của mẫu: 0.6229
Giá trị t: -6.6115
P-value: 0.0000
Bác bỏ giả thuyết H₀: Nhiệt độ cơ thể không phải là 98.6°.


### Bài 10

Tìm hiểu z_test, t_test trong python

Chọn Z-test hay T-test?

- Độ lệch chuẩn tổng thể (σ) đã biết: Sử dụng Z-test.
- Độ lệch chuẩn tổng thể (σ) không biết: Sử dụng T-test.
- Kích thước mẫu nhỏ (n < 30) và σ không biết: Sử dụng T-test.
- Kích thước mẫu lớn (n ≥ 30) và σ không biết: Có thể sử dụng Z-test hoặc T-test (T-test vẫn được ưu tiên hơn).

---